docs/proposals/user-namespace.md

```diff
@@ -0,0 +1,226 @@
+policy can check that the defaulted configuration is valid for the
+user it may be valuable and avoid frustrations with runtime errors.
+
+Pros:
+
+1. reduces the need for Linux specific knowledge
+1. system does the right thing
+
+Cons:
+
+1. requires extra flags to make the system aware of the remapped
+environment (is this required anyway due to the mount issue?)
+1. the user may unexpectedly run a container with the host user namespace.
+1. Cluster administrators must be aware that allowing certain capabilities will automatically escape user namespaces.
+
+### Option 2: API Support
+
+Fully exposing `HostUserNamespace` as a flag on the `SecurityContext` of
+a pod requires the following changes for a minimally viable feature:
+
+1. updates to the API objects
+1. support in the Kubelet to set the correct flags based on the spec
+
+In this option validation may occur in the API server if it is aware
+of the remapped environment.  This would allow rejection during the
+submission of the pod spec or, optionally, allow for the defaulting
+logic to kick in and not reject the pod.  This could be considered
+option 3.
+
+The pattern for exposing namespaces in the `SecurityContext` of a pod
+is already established.
+
+Pros:
+
+1. allows full flexibility
+1. still allows for defaulting and validation to occur (which incurs the
+cons mentioned in the defaulting option above)
+
+Cons:
+
+1. requires pod authors to have linux and cluster environment specific
+knowledge to know that the pod spec must contain a flag (unless defaulting
+is also implemented)
+
+### Option 3: Combination/staged approach
+
+As a way to begin implementation it is possible that both defaulting
+and full api support are eventually implemented.  It is useful that
+the system do the right thing or at least provide useful messaging
+to the user when misconfiguration is detected.
+
+In this option the implementation would begin with defaulting and, based
+on real world feedback, move to full API support if necessary.  Once
+API support is included a decision could be made to keep defaulting
+or start rejecting misconfigured pods in a useful way.
+
+### Kubelet
+
+In all cases, the system must ensure that the remapped UID and GID have
+access to the correct directories under `/var/lib/kubelet/pods` in order
+to perform mounting.
+
+This may be accomplished by ensuring group access to the correct
+directories.  It implies that the Kubelet will be given (or read)
+the value of `/etc/subgid` to perform a `chmod`.
+
+### Defaulting Scenarios
+
+If the system is to default host user namespace options it would occur
+when:
+
+1.  setting any other namespace flag to host.  Not strictly
+required for net=host if not binding to low ports but if the system
+provides defaulting it covers the 80% case
+1.  using capabilities that aren't tied to any namespace
+    * MKNOD
+    * SYS_MODULE
+    * SYS_TIME
+1.  explicitly granted permissions to use a host volume and escape the private user namespace
```

Comment:
This may be a use case for exposing user namespace on the security context.  If that was exposed via the API and you did not turn on defaulting you would be able to still escape user namespaces when required and then set your `--userns-remap={your managed name}` so you can manually set up the uids and groups that provide the correct DAC.  

@mrunalp @php-coder @yujuhong  thoughts here?